### ISSUE TITLE
Inconsistent handling of invalid variable values in SampleSet

### ISSUE THREAD

**Issue by randomir** [dwavesystems/dimod#529]
```
>>> import dimod
>>> dimod.__version__
'0.8.15'
>>> 
>>> bqm = dimod.BQM.from_ising({}, {'ab': 1})
>>> ss = dimod.SampleSet.from_samples_bqm([{'a': 1, 'b': 1}, {'a': -1, 'b': -20}], bqm)
>>> print(ss)
   a  b energy num_oc.
0 +1 +1    1.0       1
1 -1 -1   20.0       1
['SPIN', 2 rows, 2 samples, 2 variables]
```

Note how `-20` for `b` in the second sample is out of allowed `{1, -1}` set, yet when printed it's rounded down to `-1`. Energy stays incorrect though (calculated for `b = -20`).

**Comment by arcondello** [comment]
The energy is correct, it's calculated from `a*b*J_ab` so `-1*-20*1`. The problem is that the sampleset does not check that the input samples are actually correct so `sampleset.record.sample` in this case would be `[[1, 1], [-1, 20]]`. It is the print function, not the construction that is rounding b down.

We can change it so that the sample set enforces the samples to be the correct vartype, would you prefer an exception or rounding? There is a performance hit that comes with that, so maybe I can add a `permissive=True` flag or something.

**Comment by randomir** [comment]
Yes, I'm aware of that. I guess I forgot to write my expected behaviour:
- if invalid variable values are accepted in constructor, and actually stored in the SampleSet, then print function *should not* round them down - it should stay as close to actual sample set data as possible.
- but I don't think invalid values should *ever* be stored in the SS (since vartype is a non-optional parameter!). I like the idea of permissive flag, and I would expect that `permissive=False` raises an exception (default), and `permissive=True` rounds values.
